### ISSUE TITLE
Add type-state for Node network interfaces

### ISSUE THREAD

**Issue by quietlychris** [quietlychris/meadow#40]
A sketch of what this could look like is
```rust
use std::{fmt::Debug, marker::PhantomData};

fn main() {

    let net_cfg = NetworkConfig::<Tcp>::default();
    let node_cfg = NodeConfig::<Tcp, usize>::from_net_cfg(net_cfg);
    let node = Node::from_config(node_cfg);

    let node = Node::<Udp, Idle, usize>::default();
    let node = node.activate();
}

mod private {
    pub trait Sealed {}
    impl Sealed for crate::Udp {}
    impl Sealed for crate::Tcp {}
    
    impl Sealed for crate::Idle {}
    impl Sealed for crate::Active {}
}


pub trait Interface: private::Sealed {}
enum Tcp {}
enum Udp {}

impl Interface for Tcp {}
impl Interface for Udp {}

pub trait Message: Debug + Sync + Send + Clone {}
impl<T> Message for T where T: Debug + Sync + Send + Clone {}

#[derive(Clone, Copy)]
struct NetworkConfig<Interface> {
    __interface: PhantomData<Interface>,
    max_buffer_size: usize,
}

impl<I: Interface> NetworkConfig<I> {
    fn default() -> NetworkConfig<I> {
        Self {
            __interface: PhantomData::<I>,
            max_buffer_size: 1_000,
        }
    }
}


struct NodeConfig<Interface, T: Message> {
    __data_type: PhantomData<T>,
    network_cfg: NetworkConfig<Interface>,
}

impl<I: Interface, T: Message> NodeConfig<I, T> {
    fn default() -> NodeConfig<I, T> {
        NodeConfig {
            __data_type: PhantomData,
            network_cfg: NetworkConfig::<I>::default(),
        }
    }

    fn from_net_cfg(net_cfg: NetworkConfig<I>) -> NodeConfig<I, T> {
        NodeConfig {
            __data_type: PhantomData,
            network_cfg: net_cfg,
        }
    }
}

struct Node<Interface, State, T: Message> {
    pub __state: PhantomData<State>,
    cfg: NodeConfig<Interface, T>,
}

trait State: Default + Clone + private::Sealed {}
#[derive(Default, Clone)]
struct Idle;
#[derive(Default, Clone)]
struct Active;

impl State for Idle {}
impl State for Active {}

impl<I: Interface, T: Message> Node<I, Idle, T> {
    fn from_config(cfg: NodeConfig<I, T>) -> Node<I, Idle, T> {
        Node::<I, Idle, T> {
            __state: PhantomData::<Idle>,
            cfg: NodeConfig::<I, T> {
                __data_type: PhantomData::<T>,
                network_cfg: NetworkConfig::<I> {
                    __interface: PhantomData::<I>,
                    max_buffer_size: 1_000,
                },
            },
        }
    }

    fn default() -> Node<I, Idle, T> {
        Node::<I, Idle, T> {
            __state: PhantomData::<Idle>,
            cfg: NodeConfig::<I, T> {
                __data_type: PhantomData::<T>,
                network_cfg: NetworkConfig::<I> {
                    __interface: PhantomData::<I>,
                    max_buffer_size: 1_000,
                },
            },
        }
    }

    fn activate(self) -> Node<I, Active, T> {
        Node::<I, Active, T> {
            __state: PhantomData::<Active>,
            cfg: NodeConfig::<I, T> {
                __data_type: PhantomData::<T>,
                network_cfg: NetworkConfig::<I> {
                    __interface: PhantomData::<I>,
                    max_buffer_size: 1_000,
                },
            },
        }
    }
}
```

**Comment by quietlychris** [comment]
This also leaves open a lot of room for where the type inference occurs. For example, alternative user-facing patterns could look like 
```rust
    let net_cfg = NetworkConfig::<Tcp>::default();
    let node_cfg = NodeConfig::<_, _>::from_net_cfg(net_cfg);
    let node = Node::<_, _, usize>::from_config(node_cfg);
```

**Comment by quietlychris** [comment]
Experiments with this are at: https://github.com/quietlychris/typestates

**Comment by quietlychris** [comment]
At the moment, the current approach is to have something like `struct Node<I: Interface + Default, State, T: Message>`, which internally has `Option<TcpStream>`, `Option<Endpoint>`, and `Option<UdpSocket>` depending on what `Interface` type we're using. But what if we made `Node` a trait, and just did fully separate implementations for `Node<Tcp>` or `Node<Quic>` instead, so the `Node` wouldn't be carrying around the extra data structures around internally?

**Comment by quietlychris** [comment]
This will be addressed in the QUIC merge coming in #41. Currently, the functions available across each transport method are arbitrary; there's no functions required by the `Interface` trait or anything similar to, for example, for all Nodes to implement a `publish()` method. I think this largely makes sense, especially this earlier on, but if we are set on standardizing calls, it might be possible to create something like 
```
trait Transport {
    fn publish();
    fn request();
    fn subscribe();
}
``` 
that gets implemented for each Node type.

**Comment by quietlychris** [comment]
This was addressed in #41
